scientific-computing-with-python/learn-special-methods-by-building-a-vector-space/vectors.py

```diff
@@ -0,0 +1,97 @@
+class Vector:
+    space = 'R3'
+
+    def __init__(self, x, y, z):
+        self.x = x
+        self.y = y
+        self.z = z
+        self.norm = (x**2 + y**2 + z**2)**0.5
+
+    def __str__(self):
+        return str((self.x, self.y, self.z))
+    
+    def __repr__(self):
+        return f'Vector({self.x}, {self.y}, {self.z})'
+
+    def __add__(self, other):
+        if not isinstance(other, Vector):
+            return NotImplemented
+        x = self.x + other.x
+        y = self.y + other.y
+        z = self.z + other.z        
+        return self.__class__(x, y, z)
+    
+    def __sub__(self, other):
+        if not isinstance(other, Vector):
+            return NotImplemented
+        x = self.x - other.x
+        y = self.y - other.y
+        z = self.z - other.z
+        return Vector(x, y, z)
+    
+    def __mul__(self, other):
+        # Scalar multiplication
+        if isinstance(other, (float, int)):
+            x = self.x * other
+            y = self.y * other
+            z = self.z * other            
+            return Vector(x, y, z)
+        
+        # Scalar product
+        elif isinstance(other, Vector):
+            x = self.x * other.x
+            y = self.y * other.y
+            z = self.z * other.z
+            scalar = (x**2 + y**2 + z**2)**0.5
+            return scalar
+        else:
+            other_type = other.__class__.__name__
+            raise TypeError(f"Multiplication is not defined between 'Vector' and '{other_type}'")
```

Comment:
```suggestion
        other_type = other.__class__.__name__
        raise TypeError(f"Multiplication is not defined between 'Vector' and '{other_type}'")
```
Because there's `return` for other cases this doesn't need to be in the `else`, indented.

On the side note - is there specific reason why class raises here exception on it own, instead of letting handle it to Python, like in other cases? Or just trying to show campers various approaches?